### PR TITLE
MPI preliminaries for examples

### PR DESCRIPTION
**Pull request by FabioLuporini**
... and various changes to logging

Big diff as based on ``mpi-next`` (#629 )

### CODE REVIEW

File: devito/compiler.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 from codepy.toolchain import GCCToolchain
 
 from devito.exceptions import CompilationError
-from devito.logger import log, warning
+from devito.logger import debug, warning
```
Comment (ggorman) [inline devito/compiler.py]: Why the change from 'log' to 'debug'?

File: examples/seismic/acoustic/acoustic_example.py [hunk 2]
```diff
@@ -2,34 +2,21 @@
...
 from devito.logger import info
-from devito import Constant
+from devito import Constant, Eq, Function, Operator
 from examples.seismic.acoustic import AcousticWaveSolver
 from examples.seismic import demo_model, TimeAxis, RickerSource, Receiver
 
 
-# Velocity models
-def smooth10(vel, shape):
+def smooth(dest, src):
```
Comment (ggorman) [inline examples/seismic/acoustic/acoustic_example.py]: Calling this 'src' is *very* misleading. Call it 'kernel'.

Comment (ggorman) [inline examples/seismic/acoustic/acoustic_example.py]: Basically its hard to find anything to like about this function - to be fair the original was also also far from ideal. Why has the radius of the smoothing kernel been dropped? Why are we not using a standard Gaussian smoothing filter (ie applying principle of least astonishment)?

File: devito/yask/operator.py [hunk 3]
```diff
@@ -153,14 +153,12 @@ def apply(self, **kwargs):
...
                 toshare[v] = v.data
 
-        log("Running YASK Operator through Devito...")
         for i in self.yk_solns.values():
             i.pre_apply(toshare)
         arg_values = [args[p.name] for p in self.parameters]
         self.cfunction(*arg_values)
         for i in self.yk_solns.values():
             i.post_apply()
-        log("YASK Operator successfully run!")
```
Comment (mloubout) [inline devito/yask/operator.py]: Why removed?

File: examples/seismic/acoustic/acoustic_example.py [hunk 4]
```diff
@@ -70,8 +57,11 @@ def run(shape=(50, 50, 50), spacing=(20.0, 20.0, 20.0), tn=1000.0,
                             space_order=space_order, kernel=kernel,
                             constant=constant, **kwargs)
 
-    initial_vp = smooth10(solver.model.m.data, solver.model.shape_domain)
-    dm = np.float32(initial_vp**2 - solver.model.m.data)
+    # Smooth velocity
+    initial_vp = Function(name='ivp', grid=solver.model.grid, space_order=space_order)
```
Comment (mloubout) [inline examples/seismic/acoustic/acoustic_example.py]: would still name it `v0`, `ivp` seems weird (nitpicking)

File: examples/seismic/acoustic/acoustic_example.py (hunk 2 above)
Comment (mloubout) [inline examples/seismic/acoustic/acoustic_example.py]: `src` is confusing, we already have actual source called like that.

File: examples/seismic/model.py [hunk 5]
```diff
@@ -388,8 +388,44 @@ def initialize_function(function, data, nbpml):
...
+
+            lpad = o.left - glb_range.left if lslice == 0 else 0
+            rpad = o.right - (rslice - glb_range.right) if rslice == glb_shape[d] else 0
+
+            data_slices.append((lslice, rslice))
+            pad_widths.append((lpad, rpad))
+        except KeyError:
+            # Not a distributed dimension
+            data_slices.append((None, None))
+            pad_widths.append((0, 0))
```
Comment (mloubout) [inline examples/seismic/model.py]: Comments on what this does?

File: devito/compiler.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/compiler.py]: to avoid pollution to ``stdout``. I think the average user doesn't need to see a jit-compilation related string.

File: examples/seismic/model.py [hunk 6]
```diff
@@ -443,11 +479,6 @@ def dtype(self):
         """
         return self.grid.dtype
 
-    @property
-    def shape_domain(self):
-        """Computational shape of the model domain, with PML layers"""
-        return tuple(d + 2*self.nbpml for d in self.shape)
-
```
Comment (mloubout) [inline examples/seismic/model.py]: why is this dropped?

File: examples/seismic/acoustic/acoustic_example.py (hunk 2 above)
Comment (FabioLuporini) [reply examples/seismic/acoustic/acoustic_example.py]: the maximum radius based on the amount of halo available is automatically picked by ``src.avg``. Otherwise you (user) can pick your own, as long as you know what you're doing. 

As you said, I simply translated the original function into the Devito language, and generalised it a little bit. Nothing more nothing less

File: devito/yask/operator.py (hunk 3 above)
Comment (FabioLuporini) [reply devito/yask/operator.py]: too much stdout pollution

File: examples/seismic/acoustic/acoustic_example.py (hunk 4 above)
Comment (FabioLuporini) [reply examples/seismic/acoustic/acoustic_example.py]: it was called ``initial_vp``, happy to change it to v0 !

File: examples/seismic/model.py (hunk 5 above)
Comment (FabioLuporini) [reply examples/seismic/model.py]: Yes, I can elaborate with a better comment. Basically, it takes the entire array, as there's nothing to be distribtued along the given dimension

File: examples/seismic/model.py (hunk 6 above)
Comment (FabioLuporini) [reply examples/seismic/model.py]: now unused

File: examples/seismic/acoustic/acoustic_example.py (hunk 4 above)
Comment (FabioLuporini) [reply examples/seismic/acoustic/acoustic_example.py]: done

**Comment by FabioLuporini** [comment]
Merged.
